activejob/lib/active_job/enqueue_after_transaction_commit.rb

```diff
@@ -0,0 +1,51 @@
+# frozen_string_literal: true
+
+module ActiveJob
+  module EnqueueAfterTransactionCommit # :nodoc:
+    extend ActiveSupport::Concern
+
+    included do
+      ##
+      # :singleton-method:
+      #
+      # Defines if enqueueing this job from inside an Active Record transaction
+      # automatically defers the enqueue to after the transaction commit.
+      #
+      # It can be set on a per job basis:
+      #  - `:always` forces the job to be deferred.
+      #  - `:never` forces the job to be queueed immediately
+      #  - `:auto` uses the queue adapter default behavior (recommended).
+      class_attribute :enqueue_after_transaction_commit, instance_accessor: false, instance_predicate: false, default: :never
+
+      around_enqueue do |job, block|
+        after_transaction = case job.class.enqueue_after_transaction_commit
+        when :always
+          true
+        when :never
+          false
+        else # :auto
+          queue_adapter.respond_to?(:enqueue_after_transaction_commit?) && queue_adapter.enqueue_after_transaction_commit?
+        end
+
+        if after_transaction && defined?(::ActiveRecord::Base)
+          ActiveRecord::Base.after_current_transaction_commit(&block)
+        else
+          block.call
+        end
+      end
+    end
+
+    module ClassMethods
+      def assign_adapter(adapter_name, queue_adapter)
+        super
+
+        unless queue_adapter.respond_to?(:enqueue_after_transaction_commit?)
+          class_name = queue_adapter.is_a?(Module) ? queue_adapter.name : queue_adapter.class.name
+          ActiveJob.deprecator.warn(<<~MSG)
+            #{class_name} doesn't implement the `#enqueue_after_transaction_commit?` method. Assuming false.
```

Comment:
My instinct is that we can silently assume `true` (so without even producing a warning), and then just allow DB-focused adapters to opt in to the `false` mode purely as an optimisation: given that the contract is always to do the work at some moment later than right-now, I think deferring the enqueue is not a change in contracted behaviour.

Separately, and for lack of a better place for this observation: it's possible that a boolean `enqueue_after_..` isn't specific enough. As is, the default behaviour would be to wait for all DBs' transactions to commit for an `after=true` adapter, but to only wait for the _queue_ DB's transaction when `after=false`. That feels like a dangerous trap for people moving between adapters, but I don't yet have a good suggestion of what we'd do instead. 🤔

Reply:
> to only wait for the _queue_ DB's transaction

Yeah, that's something I thought of a bit, and I think I'm "ok" with this limitation, as you really need quite a few conditions to fall into it:

  - You need to use a DB backed adapter.
  - Have a multi-db setup.
  - Have one of these DB be used for both jobs and other things.

Because the idea is that if you use something like SolidQueue with its own dedicated database, then you should configure it to enqueue on after commit like Sidekiq or Resque would. That is why I added an `initialize` parameter to some adapters.

If we really wanted to handle this edge case at all cost, we'd need the queue_adapter to tell us which connection it uses (e.g. return an AR connection object) so we can check if that's the only one with an open transaction. 

It's not that hard to implement, but the coupling would be quite nasty. All this to say I considered it, but thought it wasn't quite worth it.

Reply:
> My instinct is that we can silently assume `true`

Right, I hesitated a bit on that one. I'm totally OK with assuming `true`, my only worry is the discoverability for people implementing adapters. But I suppose it's fine.

Another reason was to be able to remove that `respond_to?` in the future, but I can live with it.

Reply:
If we don't need to check whether it's been manually defined for the warning, I think we can just shove a default implementation in there, and avoid the `respond_to?` that way?

Reply:
> Because the idea is that if you use something like SolidQueue with its own dedicated database, then you should configure it to enqueue on after commit

Yeah.. I don't love that that leaves one of the most straightforward configurations of those adapters (indeed the configuration that I suspect we might recommend, because of the different data/throughput shapes involved) as needing explicit "please behave like [the new] normal" instructions.

> It's not that hard to implement

The bit where I trailed off on my mental "we could just.." was what we'd do if the queue DB has an open transaction, then that transaction wants to commit, but we want to hold for some other separate transaction: despite my "we can always queue work later" argument, pushing the enqueue out of an explicit transaction feels like A Choice. :confused:

Reply:
> whether that would be worth the churn.

I think it would be fairly simple. Rails itself maintains the major adapters, and the more recent adapters (GoodJob, Solid Queue) are actively maintained. 

I think my only question is how ugly it would be for my adapter to maintain backwards compatibility (I imagine I'd just do something like `class ActiveJob::BaseAdapter; end unless defined?(ActiveJob::BaseAdapter)` but hopefully you'd tell me the best thing to do)

Reply:
My only opposition to this is that it might create a bit of lag on the upgrade as users of 3rd party adapters will have to wait for a gem upgrade.

But as you said, the most common ones are included anyway, so maybe it's fine.